versions/1.1/SPEC.md

```diff
@@ -0,0 +1,5771 @@
+
+* B (bytes)
+* Decimal: KB, MB, GB, TB
+* Binary: KiB, MiB, GiB, TiB
+
+An optional space is allowed between the number/expression and the suffix. For example: `6.2 GB`, `5MB`, `"${ram}GiB"`.
+
+The decimal and binary units may be shortened by omitting the trailing "B". For example, "K" and "KB" are both interpreted as "kilobytes".
+
+#### Mandatory `runtime` attributes
+
+The following attributes must be supported by the execution engine. The value for each of these attributes must be defined - if it is not specified by the user, then it must be set to the specified default value. 
+
+##### `container`
+
+* Accepted types:
+    * `String`: A single container URI.
+    * `Array[String]`: An array of container URIs.
+
+The `container` key accepts a URI string that describes a location where the execution engine can attempt to retrieve a container image to execute the task.
+
+The user is strongly suggested to specify a `container` for every task. There is no default value for `container`. If `container` is not specified, the execution behavior is determined by the execution engine. Typically, the task is simply executed in the host environment. 
+
+🗑 The ability to omit `container` is deprecated. In WDL 2.0, `container` will be required.
+
+The format of a container URI string is `protocol://location`, where protocol is one of the protocols supported by the execution engine. Execution engines must, at a minimum, support the `docker://` protocol, and if no protocol is specified, it is assumed to be `docker://`. An execution engine should ignore any URI with a protocol it does not support.
+
+```wdl
+task single_image_test {
+  #....
+  runtime {
+    container: "ubuntu:latest"
+  }
+```
+
+Container source locations should use the syntax defined by the individual container repository. For example an image defined as `ubuntu:latest` would conventionally refer a docker image living on `DockerHub`, while an image defined as `quay.io/bitnami/python` would refer to a `quay.io` repository.
+
+The `container` key also accepts an array of URI strings. All of the locations must point to images that are equivalent, i.e. they must always produce the same final results when the task is run with the same inputs. It is the responsibility of the execution engine to define the specific image sources it supports, and to determine which image is the "best" one to use at runtime. The ordering of the array does not imply any implicit preference or ordering of the containers. All images are expected to be the same, and therefore any choice would be equally valid. Defining multiple images enables greater portability across a broad range of execution environments.
```

Comment:
I'd prefer the execution engine here to simply try them in order and use the first one that works:

+ Simpler logic in the execution engine
+ Predictable behaviour.
+ Simpler spec

Reply:
Non-content comment: This is one of the examples of way to many sentences on one line. Maybe use one line per sentence here.

Reply:
Right, but that is not what is stated in @patmagee PR #315. The discussion should be had there.

Reply:
My answer from slack
> my intention was not to give an equivalent of CSS font selection (ie fallback options if the main container is not available). Containers should all be the exact same (ie same hash) and the engine can choose whatever is most optimal

The reasoning for this was that certain container repos would be supported by an engine while others were not. So in this scenario the engine can choose whichever container it likes, and still is able to run the task in the exact same way. 

Take the following example:
ie: 
```wdl
runtime {
  container: ["foo/image:latest", "bar/image:latest", "bizz/image:lateast"]
}
```

Does not imply the precedence relationship of `"foo/image:latest" > "bar/image:latest" > "bizz/image:latest"` but the equality relationship of `"foo/image:latest" == "bar/image:latest" == "bizz/image:latest"` (which ideally would be provable by the engine, but is unlikely if different engines simply cannot read for a different repository)

Reply:
Ah okay. Well since this was already discussed elsewhere so no need for me to challenge this decision. If this solves problems for people I am happy it is there.